### DISCUSSION TITLE
Improving Symbol Names

### DISCUSSION THREAD

**Discussion by pbackus** [dlang/PhobosV3-Design#15] [General]
Which symbols in Phobos are poorly-named, and how might they be improved?

**Comment by pbackus** [comment]
`countUntil` should be renamed to `indexOf`.

A forum [search for "indexOf countUntil"][1] reveals just how many people have gotten caught in this trap over the years.

The full history of the name is recounted in [this post by @schveiguy][2]. 

[1]: https://forum.dlang.org/search?q=indexOf%20countUntil
[2]: https://forum.dlang.org/post/qugcta$112n$1@digitalmars.com

**Reply by crazymonkyyy** [reply]
countUntil is correct given the "ranges are views of data" world view and using countUntil for an index is just wrong 
instead there should be index generated by the datastucutres or .asUnicode, data strucute aware functions

and then allot of searching should be replaced with functions like indexes(alias F)()=>filter!F.swapKeyValue

**Reply by pbackus** [reply]
`countUntil` isn't wrong. It's just that `indexOf` is better and more intuitive.

In fact, if you read the linked post, you'll see that `countUntil` was originally named `indexOf`.

**Reply by crazymonkyyy** [reply]
countUntil is the correct name for what the function does
its just the wrong function, if a user wants indexes there should be an array or hash type underneath 

`foo.retro.countUntil` `foo.toUnicode.countUntil` `foo.filter.countUntil` are all, doing exactly what its saying, but its very fragile

**Reply by schveiguy** [reply]
I somewhat agree with @crazymonkyyy here. indexOf implies it returns an index, which can be used to look up the value found.

I would propose that indexOf is an alias to countUntil, but only for types that have indexing. This should fix most of the confusion.

**Reply by crazymonkyyy** [reply]
are you supporting my `index` as part of the api or impling that there should be algorthims that act on datastuctures?

**Reply by schveiguy** [reply]
I'm saying `indexOf` seems odd for a type that has no indexing. Like `filter`. But it makes perfect sense for ones that do, and most people know `indexOf` from other languages/experience, so it makes sense to add it back for random-access ranges.

**Reply by crazymonkyyy** [reply]
whats should happen for 
`"foo :smile: bar".asUnicode.indexOf('b');`
`iota(10).drop(3).indexOf!(a=>a%2);`

searching is primarily about strings, and there are random access ranges that are very abstract and not about data that literally exists(and therefore has an index for you to assign to)

**Reply by pbackus** [reply]
I think the existing empirical evidence shows very clearly that `indexOf` is the more intuitive name, even in the case where the range doesn't support indexing. If you're not convinced, though, let's gather more data (e.g., run a poll on the forums) rather than just sitting here in our armchairs debating.

**Reply by schveiguy** [reply]
Consider the case with phobos2 (current phobos), and autodecoding.

```d
assert("héllo".countUntil('o') == 4);
assert("héllo"[4] == 'l');
```

Now, change `countUntil` to `indexOf`, and it's going to be a real head-scratcher.

Granted, phobos v3 will remove autodecoding, but the fact remains, the return value is *not* an index. It's a count.

`indexOf` is intuitive because you are already used to it from every other language. But in those languages, indexOf is a method on a *type*, not a general algorithm. And of course, it only works where defined. Which is why I said, let's define it for places where you actually have indexes.

Looking at C++, they take the much much worse route, as you have to do:

```c++
std::distance(begin(), std::find(begin(), end(), val));
```

or use `boost::counting_iterator` or something like that.

**Reply by crazymonkyyy** [reply]
>  empirical evidence

how do you break down the complaints between about the current situation being wrong because the name is bad vs the function itself being bad empirically?

Everyone agrees that the current situation where strings get autodecoding and that the std has bad names; im pushing it futher that the function *itself* illuminates the "ranges as views" is a problem as you can see from everyone want to instantly use the function incorrectly, wanting to have the function be different name that reflects how its used, even if that use likely wrong

**Reply by crazymonkyyy** [reply]
> let's define it for places where you actually have indexes.

the primary use case is strings; and string probaly wont have indexes if people continue with stl inspired version of iteration
its not by chance std.string has a bunch of searching functions; and if theres an attempt to cut down on specail treatment on strings, without making ranges fundamentally compatible with unicode they will just get copy and pasted back

**Reply by pbackus** [reply]
> the return value is not an index. It's a count.

What's the difference? The index of an element in a range is equivalent to the count of the elements that come before it. I don't see how there's any confusion or ambiguity there.

**Reply by crazymonkyyy** [reply]
counts are not indexs for chunk, filter, unicode handling, retro, drop, splitwhen+joiner, chain, cycle, repeat, trasposed

its only vaguely naturally correct for map, zip and slide

**Reply by schveiguy** [reply]
Indexes are indexes, not counts. An index is a token that can be used to look up a specific element. Sometimes they happen to be equivalent to counts, sometimes not.

```d
int[int] aa = [1:1, 2:2, 3:3];
auto idx = aa.byValue.indexOf(3);
assert(aa[idx] == 3);
```

**Reply by pbackus** [reply]
@crazymonkyyy Counts and indexes are always equivalent with respect to any given range.

Obviously a count/index in one range is not always going to be the same as the count/index in a different range, or in the underlying container from which a range is derived. I think most programmers understand this intuitively. For example, if asked what the following code snippet would print:

```d
writeln([1, 2, 3, 4, 5, 6].filter!isEven.indexOf(6));
```

...how many programmers do you think would honestly answer `5` instead of `2`?

Again, if you disagree, let's go and gather some actual data.

**Reply by pbackus** [reply]
@schveiguy Your AA example only works because you're conflating `aa.byValue` and `aa` itself, just like the autodecoding example (in effect) conflates `"héllo".byDchar` with `"héllo"` itself. Obviously if you use an index from one thing to look up an element of a different thing, you will get nonsense.

If we correct the example, like this:

```d
int[int] aa = [1:1, 2:2, 3:3]
auto idx = aa.byValue.indexOf(3);
assert(aa.byValue[idx] == 3);
```

...then there's no problem.

**Reply by crazymonkyyy** [reply]
I dont disagree that peoples expectation; im suggesting they *want* 5 as seen from strings; or if Im using `retro.indexOf` ill always be risking an off by one error when I call foo[$-i-1]

every enumerate.filter was a failure of api and id bet 90% of enumerates are followed by one of the functions I listed along with a bunch of "just work" code

**Reply by schveiguy** [reply]
> ...then there's no problem.

Except that doesn't compile -- the range is not a random-access range. So you can imagine that one thinks "hm... it gave me an index, that must index *something*. Oh, this works!"

**Reply by pbackus** [reply]
> So you can imagine that one thinks "hm... it gave me an index, that must index *something.*"

Do you think this? Have you personally spoken to someone who thinks this? Can you point me to any forum post, Discord message, or other written record of someone thinking this?

I've never encountered anyone who thinks this. But I've encountered a lot of people who are confused about why `countUntil` isn't named `indexOf`.

By the way, the same objection applies to `foreach`:

```d
int[int] aa = ...;
foreach (i, val; aa.byVal) {
    doSomethingWith(aa.byVal[i]); // doesn't compile
}
```

I don't think anyone has ever complained that `foreach`-with-index is too confusing, because sometimes you can't use the index to look up the range element.

**Reply by schveiguy** [reply]
I'd have the same objection if the name was called `pointerTo` that didn't return a pointer. Names should say what they do, not what they don't do. If something says `indexOf`, it should return the index of something. To me, the term "index" means a specific thing, and the count of things that came before the item isn't it. For an array, these terms happen to coincide. For many other things, they don't.

But this argument is kind of silly. I *agree* `indexOf` should be there for all the places where people are expecting it (i.e. for an indexable thing). It's a much better name for finding indexes than `countUntil`. For things without indexing, `countUntil` isn't my favorite name, but it already exists. If there was a better name, I'd be for that.

BTW, the foreach statement `foreach(i, val; aa.byValue)` does not compile -- `byValue` isn't an index-supporting range.

**Reply by schveiguy** [reply]
Let's take an example of a linked list. This is something that really isn't indexable in the traditional sense.

Java's LinkedList has an `indexOf` operation. However, it also has a `get(int index)` method! What do these do? As you expect, iterate from the front or back until it gets to the object. So even though it really shouldn't exist, it does. That is a design decision that Sun made where they thought it was more important to provide the same interface for all List types.

.Net's LinkedList does *not* contain an `IndexOf` method, and accordingly, it doesn't have an operator for accessing values by index.

Even though the two libraries differ on whether linked lists should be indexable, they are consistent in presenting the interface: No indexing -- no `indexOf`. This is expected behavior, and should not confuse anyone who is used to `indexOf` from other languages.

**Reply by pbackus** [reply]
To me, "index" means "a number denoting the position of an element in a sequence." It doesn't necessarily imply that you can look up the item directly given the index.

For example, I could say something like "let *p<sub>i</sub>* be the *i*-th digit of pi after the decimal point." In this context, I would be comfortable referring to *i* as an "index", even though there is no direct way to look up *p<sub>i</sub>* for a given value of *i*.

Similarly, we can say "let *e<sub>i</sub>* be the *i*-th element of an input range *r*", and refer to *i* as the index of *e<sub>i</sub>*, even though there may be no direct way to look up *e<sub>i</sub>* given *i* and *r*.

I think this meaning of "index" is widely-used enough, and easy enough to understand, that D programmers are unlikely to be confused by it, even if it is not the first meaning that comes to mind when they hear the word.

**Reply by pbackus** [reply]
Another possibility is the name used in Rust: [`position`][1].

[1]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.position

**Reply by crazymonkyyy** [reply]
in d its called opIndex when you make an array or hashmap
ill look forward to your opPos dip for better clarity and mimiking rust

**Reply by schveiguy** [reply]
> To me, "index" means "a number denoting the position of an element in a sequence." It doesn't necessarily imply that you can look up the item directly given the index.

A reasonable position. But it's not my position, and I also think D does not have that view. Maybe I'm wrong though?

> Another possibility is the name used in Rust: [position](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.position).

They basically have the same terminology: "returning its index". And I see that Rust iterators also have a "nth" operation that looks up a value by "index", regardless of the cost of that operation. It also is a destructive method, advancing the iterator. It seems like we have something like that with `dropExactly(count).front`.

I can concede that both points of view are defensible. I do prefer the distinction between indexable things and non-indexable things based on algorithmic complexity, to avoid accidentally doing linear lookups when you think it's quick.

**Reply by pbackus** [reply]
> I also think D does not have that view

Examples of "index" used in the broad sense in Phobos:

* [`enumerate`][1]
* [`minIndex`][2] and [`maxIndex`][3]

The template constraints for these functions don't require their inputs to support indexing.

You can argue that these are mistakes if you like, but I've never heard of anyone being confused by them.

[1]: https://dlang.org/phobos/std_range.html#enumerate
[2]: https://dlang.org/phobos/std_algorithm_searching.html#.minIndex
[3]: https://dlang.org/phobos/std_algorithm_searching.html#.maxIndex

> I do prefer the distinction between indexable things and non-indexable things based on algorithmic complexity, to avoid accidentally doing linear lookups when you think it's quick.

Given that range-based code can and does already check for indexability directly (with `isRandomAccessRange`), I don't think this distinction is in any danger of being lost, even if we decide to accept broader use of the term "index." I do agree that the distinction is important.

**Reply by rikkimax** [reply]
I suspect that we should review this after Walter has done his talk on matching.

If possible a sentinal based sumtype in library would allow us to properly represent an index via the use of ``ptrdiff_t`` and anything that did not use this, shouldn't be considered an index.

**Reply by pbackus** [reply]
@rikkimax We can express this already with `Nullable!(size_t, size_t.max)`.

Either way, this dicussion is about naming, not representation of indices. Please feel free to start a new discussion about the use of nullable/optional/sum types in Phobos interfaces if you have thoughts on that.

**Reply by rikkimax** [reply]
What I am suggesting is that one implies the other as a convention.

Being consistent with our names as to what they imply in the signature as to what is supported is important.

**Reply by pbackus** [reply]
@rikkimax We can use hyperlinks to connect related discussions if desired. Let's keep this discussion on-topic.

**Reply by schveiguy** [reply]
OK, so we should be clearer that "indexing" does not mean using an "index", it means using a "key" (which could be an integer, and could coincide with the index of the value), and an index is specifically a serial number which is ever incrementing.

I still think indexOf on arrays and other random-access ranges would solve 99% of the problems.

**Reply by crazymonkyyy** [reply]
its an index for arrays and key for hashmaps; but arrays > hashmaps and on some level index is verbose and people know what an 'i' is

**Reply by rikkimax** [reply]
What you have described with a number that increments, is what I would call in my stuff an offset.

An index is a ``ptrdiff_t`` value that may be negative to refer to decremented from end.
You must convert the index to an offset before usage.

For things like maps, a key is a key, not an index.

This is why we have to get the types down, there are differing opinions, and for consistency, if we can define names against behavior it is beneficial.

**Reply by crazymonkyyy** [reply]
> This is why we have to get the types down, there are differing opinions, and for consistency, if we can define names against behavior it is beneficial.

can't say I havn't been trying

https://github.com/LightBender/PhobosV3-Design/discussions/21
https://github.com/LightBender/PhobosV3-Design/discussions/17
https://github.com/LightBender/PhobosV3-Design/discussions/13

```d
auto indexof(alias F,T)(T array){
  foreach(i,ref a;array){
    if(F(a)){
      assert(&a==&array[i]);
      return i;
 }}}
 ```
 because of functional programming got most of the data structures and algorthims work, we picked up the "views of data" thing, but thats not what I mean by an index, I mean an 'i' in a for loop

**Reply by rikkimax** [reply]
Inside the foreach loop is easy, that is an offset.

What happens if it doesn't find it is the problem.
That needs defining and in turn defines what the name should contain.

**Reply by crazymonkyyy** [reply]
in what way is it an offset when I pass a `string[string]` to my version of index of? That 'i' is short for either index no?

**Reply by rikkimax** [reply]
I would suggest:

An index is a ``ptrdiff_t`` sequential value that may be boxed, negative values are added to the length.
For searching algorithms a ``-1`` value will indicate not found, otherwise it is an offset.
All values after absoluting are between ``0`` and ``length``.

An offset is a ``size_t`` value that is sequential.
All values are between ``0`` and ``length``.

A key is anything that refers to another that does not match an index or offset.

If any rule matches for searching or retrieval they should use the name of data representation.

**Reply by rikkimax** [reply]
You shouldn't be searching a map like you did.

The result value could be different every time you call it.

This should be something completely different on the data structure itself.

**Reply by crazymonkyyy** [reply]
I don't think you should want specif types in the range api; I have strong opinions of what the right index for a 2darray is, a tuple!(int,int), and that won't match a ptrdiff_t

I can't stress enough how much I think a range.index should be a value thats ussally generated by an array or hash type that will point you to real data; if it doesnt have a .index its a stream/virtual if it does it should fit into a opIndex somewhere and what that *means* is the users responiblity

**Reply by rikkimax** [reply]
```d
import std;
void main()
{
    int[] array = [1, 2, 3];
    ptrdiff_t index = 1;
    writeln(array[index]);
}
```

``ptrdiff_t`` works fine here when it is a positive number for lookup into an array.

Negative values would be for lookup and are quite convenient.

Note: this definition is what I am now using and it does work.

**Reply by crazymonkyyy** [reply]
in what way is that a 2d array or an aa?

**Reply by rikkimax** [reply]
D does not have 2d arrays. We only have 1 dimensional arrays.

It is not a map and hence not an AA.

As a result I don't understand what you are asking.

**Reply by crazymonkyyy** [reply]
I think a 2d array was the least converal thing Ive asked for and is definitely part of snar's data structure list
ranges should be able to work with nd arrays and an nd arrays index will be a user-defined tuple

**Reply by pbackus** [reply]
I think "offset" would be a bad choice, because it usually implies a contiguous memory layout.

**Reply by rikkimax** [reply]
That wouldn't work with a generic search algorithm.

You need to be able to choose how a search takes place.

I.e. is it ``[1, 0]`` then ``[1, 1]`` or is it ``[2, 0]``?

Paul: I would suggest for search algorithms that they should be an index based upon my above rules, not an offset due to the possibility of the value not being found.

**Reply by crazymonkyyy** [reply]
> I.e. is it [1, 0] then [1, 1] or is it [2, 0]?

Are you imagining that indexOf would construct the indexs?
```d
auto indexOf(alias F,R)(R r) if(monkyyrange!R){
  while( ! F(r.front)){
    r=r.pop;
  }
  return r.index;
}
```
if a nd array range defines a .index that fits into its opIndex, then it just works

**Reply by rikkimax** [reply]
It may compile, but it may not give the result someone needs.

Sequentiality has to be defined to fit a person's use case.

And this is a key concept of what I described with offset and index, both of these have a natural order based upon the sequential nature of whole numbers.

Multiple dimensional arrays can have multiple sequences that are all valid, this makes them a key, not an index or offset.

**Reply by crazymonkyyy** [reply]
> It may compile, but it may not give the result someone needs.
> natural order

Its *already* the datastructures responsibility to define ranges in the natural ways to iterate

a tree will need both a depth first and breath first iteration, and if you think a 2d array needs a (always slower) colum based iteration, I geuss you should tell snar to add it to his todo list 

tree.depthfirst.indexOf and tree.breathefirst.indexOf should be expected natural code

**Reply by jmdavis** [reply]
I agree we should not be conflating indexing and counting. That's slightly less critical when auto-decoding isn't a thing, but they're distinct concepts which imply completely different algorithmic complexity, and I think that we should keep them that way. And if there are places where we screwed that up in Phobos v2, then we should fix it in v3, not compound it.

**Reply by pbackus** [reply]
Just got another question in the community Discord where someone was asking for `indexOf`, and I had to point them to `countUntil`.

**Reply by LightBender** [reply]
IMO, this one has to change. The rest of the world uses different terms than we do for it. That's why one of the naming rules is to do an audit of the StdLib's of major languages to see what they did.

**Comment by pbackus** [comment]
Prefixes should be dropped from all names in `std.meta`.

Currently, three different styles of name are used in `std.meta`:

1. Prefixed with `static` (e.g., `staticMap`)
2. Prefixed with `template` (e.g., `templateOr`)
3. Capitalized with no prefix (e.g., `Filter`)

Most names currently use style (3). The remaining prefixed names should be changed to match.

**Reply by jmdavis** [reply]
All that's been ported in that camp thus far is `staticMap`, but I went with `Map`, which matches the Phobos naming convention for types.

Historically, I think that the problem came about, because while `AliasSeq`s _usually_ contain types in the context of std.meta, they don't always, so there was debate over whether the naming convention for types should be followed, and we got a mess as a result.

However, they are almost always types, and the result is shorter, so IMHO, it makes sense to just go with #3.

**Reply by ntrel** [reply]
Yes, if it might hold a type then it should be capitalized. This is why `AliasSeq` is capitalized.

**Comment by crazymonkyyy** [comment]
all io naming conflicts should be resolved
dirEnties seems bad; and probably should be replaced with a "pathRange" and some sane defaults wrappers of "iterate over files in this folder", "all .jpg files", "~/.config" or "%appdata$/" depending on os

**Comment by pbackus** [comment]
[`Flag!"each"`][1] should be given a more descriptive name.

Currently, the callback passed to `each` may return `Yes.each` to continue iteration, or `No.each` to stop. This name is not meaningful, and makes it hard to tell which value corresponds to which behavior.

The ideal name for this flag would be `Flag!"continue"`, but `Yes.continue` and `No.continue` are syntactically invalid, because `continue` is a keyword.

[1]: https://dlang.org/phobos/std_algorithm_iteration.html#each

**Reply by schveiguy** [reply]
I thought one of the main points of named parameters was to get rid of `Flag`?

**Reply by pbackus** [reply]
@schveiguy `Flag!"each"` is a return value, not a parameter.

**Reply by schveiguy** [reply]
Ok, I see, in this specific context, I was thinking about `Flag` in general.

that is... an odd one. I'd say just return a specific enum? The use of `Flag` is a code smell in any case.

But also note, that `Yes.each` means something different when `each` returns, which makes it even worse (whether it iterated the entire thing or not). It's being overloaded to mean 2 different things.

**Reply by jmdavis** [reply]
Honestly, I'm very tempted to argue that we should get rid of `each` entirely. All it is is a bunch of code to not have to explicitly use `foreach`. I do not understand why it's even a thing. It's a bunch of complicated code just to turn

    foreach(e; range)
    {
        ...
    }

into

    range.each!(e => ...)();

and in my experience, the result tends to be harder to read / understand to boot - especially if the lambda provided is complex.

**Reply by crazymonkyyy** [reply]
range.each!writeln is fully anoumous, has no ()'s; its not useless

**Reply by jmdavis** [reply]
When you could have just written

    foreach(e; range)
        writeln(e);

and avoided needing a 200 or so line template function in Phobos, would have instantiated fewer templates, and would have gotten a slightly faster compile time. All it does is move the body of a loop into a template argument. I really don't understand why that's worth having.

**Reply by crazymonkyyy** [reply]
phoboes being verbose and overcomplicated does not justify removing api
I'd aim for a 5-10 line `each`

**Reply by rikkimax** [reply]
I've used it in the past once you start getting a bit of depth on your ranges.

Now don't get me wrong, that is arguable if you should be doing it.

But people will want it, they will request it so we might as well have it.

**Reply by jmdavis** [reply]
I would definitely argue that `each` does so little that it's not worth having in Phobos v3 (especially with how much code it is), so if I were making the decision, it would not make it in, but I also don't feel so strongly about it that I'm going to go on the warpath over it. If other folks want it to make it into v3, presumably it will.

I truly don't understand why it's even a thing, but it's also not fundamentally broken or anything. It's just a bizarre amount of code just to avoid an explicit loop. It rearranges code without actually enabling anything, making it pointless IMHO.

**Reply by rikkimax** [reply]
Once you have depth it isn't possible to rewrite it as a foreach since it's a statement, not an expression taking in an argument.

Consider a map whose element type is an array.
You process each element as its own range.
This is depth.

``.map!((v) { v[0].thing.each; v[1].thingie.each; }``

Except it's not all on one line and it's much longer.

**Reply by jmdavis** [reply]
which IMHO results in a convoluted mess in comparison to `foreach`. And since `each` is consuming the range, it doesn't make a lot of sense to use it in `map` like that IMHO. I'd just use `foreach` on the main range and then `foreach`es on each of the elements, making for much more legible code. Any time I've seen anyone try to use `each` with anything that isn't dead simple, it's been a convoluted mess where a `foreach` would have been much easier to read.

But whatever. As I said, I don't see any value in it, but it's not so terrible that I'm going to fight for it to not make it into v3.

**Reply by schveiguy** [reply]
I'm kind of with @jmdavis. What is wrong with just having [`std.range.tee`](https://dlang.org/phobos/std_range.html#tee) for this? tee is more useful, as it can be in the middle of a range instead of the end of it.

I've never used `each`.

**Reply by rikkimax** [reply]
That is done lazily.

If you don't iterate over the tee, the elements will never be seen by the function.

I've used each in places where this would never be called.

**Reply by schveiguy** [reply]
Then you use `count` on it.

**Reply by rikkimax** [reply]
Okay that is at least close to what ``each`` offers.

It'll be hard for people to find however, it was never something I even considered for this and this alone goes against the new objective of using names that people expect.

**Reply by schveiguy** [reply]
We should be using building blocks. `count` (or `walkLength`) is the closest thing to "iterate this whole range" that we have. Something like `void iterate(Range)` would work too, but no reason not to return  *something* like number of things iterated.

But my impression is that `each` conflates calling the function with evaluation. `tee` focuses on calling a function on each of the elements, and it's a building block which you can then evaluate (by using foreach, or by calling `count`). If you want to call multiple functions, you can chain them together with multiple `tee`, giving even more of a feel of functional programming. `each` always ends the chain of calls.

Note, I have a similar feature in iopipe: https://schveiguy.github.io/iopipe/iopipe/bufpipe/process.html

**Reply by pbackus** [reply]
Suggesting that D programmers should write `range.tee!fun.count` instead of `range.each!fun` is like suggesting that shell users should write `tee /dev/null < file` instead of `cat file`. I can't take it seriously.

I agree that the current implementation of `each` has a lot of unnecessary complexity, and would support replacing it with a simpler implementation.

**Reply by pbackus** [reply]
If we *really* want to get rid of `each`, my proposed alternative would be a `sink` template that takes a callback and creates an output range from it. Then you could write:

```d
someRange
    .doWhatever
    .copy(sink!fun); // instead of each!fun
```

With an improved output range interface (`bool put(E e)`, which returns `false` on failure), this can even support early termination.

**Reply by rikkimax** [reply]
It would achieve the goal of ``each`` from what I have needed in the past, therefore no opinion from me.

**Reply by crazymonkyyy** [reply]
I use each when map doesnt work with void functions; I assumed theres a mathy style reason map doesn't have a void element type overload that's eager
if each was merged into map Id perfer that

**Reply by crazymonkyyy** [reply]
> We should be using building blocks. count

I think a "last" would be a better building block 
"iterate a range, until you find the end,(backtrack 1) return last valid state" 
`filter.last.index` `cumitivesum.filter(not nan).last`

**Reply by ArthaTi** [reply]
`all` and `any` already do this. 

```
only(1, 2, 3)
    .all!((a) {
        writeln(a);  // prints 1 and 2
        return a < 2;
    });
```

It's not an obvious solution, but it is one.

**Comment by crazymonkyyy** [comment]
allot of the meta range function got overly latinized names 
iota instead of counter
retro instead of backwards
Transversal instead of takeColume

**Reply by ntrel** [reply]
`counter` would be a *much* better name than `iota`. *iota* definition is "an extremely small amount". Also `3 in counter(0, 5)` makes far more sense, `iota` doesn't make sense there at all. I knew Phobos supported that expression, but just couldn't remember the function name (after a few days on the NG, @pbackus [pointed out](https://forum.dlang.org/post/polxndivboyeucmrvfyp@forum.dlang.org) that `iota` supports `in`).

**Reply by pbackus** [reply]
For the record: the name "iota" comes from [`std::iota` in C++][1], which in turn comes from [the iota operator in APL][2].

I agree that it is a bad name and should be changed in Phobos v3.

[1]: https://en.cppreference.com/w/cpp/algorithm/iota.html
[2]: https://docs.dyalog.com/20.0/language-reference-guide/primitive-functions/index-generator/

**Comment by pbackus** [comment]
`std.algorithm.iteration.group` should be made into an overload of `uniq`.

In Unix terms, what `group` does is analogous to `uniq -c`. In the same way that the `uniq -c` command is a variant of the `uniq` command, the `group` function should be a variant of the `uniq` function.

**Comment by crazymonkyyy** [comment]
several traits seems strange and should be combined, like I have to look up perimeters vs fields, instead of using fields for both functions and structs to get a tuple of types

**Comment by pbackus** [comment]
`std.random.dice` should be given a more descriptive name, or removed.

This function has nothing to do with dice. It samples from what Wikipedia calls a ["categorical distribution"][1], which is a special case of a [multinomial distribution][2]. The result of rolling a die, by contrast, has a uniform distribution.

[1]: https://en.wikipedia.org/wiki/Categorical_distribution
[2]: https://en.wikipedia.org/wiki/Multinomial_distribution

**Reply by ichordev** [reply]
I agree with giving it a more descriptive name. There's a note in `std.random` with a good suggestion:

> In future Phobos versions, this function might get renamed to something like `weightedChoice` to avoid confusion.

**Reply by Ogi-kun** [reply]
I think other `std.random` functions could also use more descriptive names.
`uniform` → `randomUniform`, 
`uniform01` → `randomUniform01`, 
`uniformDistribution` → `randomUniformDistribution`,
`choice` → `randomChoice` (This one especially. I confuse `choice` and `choose` all the time), 
`dice` → `randomChoiceWeighted`, 
`randomCover` → `asRandomShuffled` (like e.g. `asUpperCase`).

Maybe `uniformDistribution` should be removed entirely, given that it’s not commonly used and is trivial to implement.

**Reply by pbackus** [reply]
In general I think we should avoid repeating the module name in the names of individual functions.

My suggestions:

* `randomSample` → `sample`
* `choice` → `sampleOne` (like `takeOne` and `dropOne` from `std.range`)
* `randomCover` → `shuffled`
* `randomShuffle` → `shuffle`

**Comment by ntrel** [comment]
`std.traits` should not use `Tuple` to mean a sequence. Instead I suggest we rename these templates as per https://github.com/dlang/phobos/pull/6227.

**Comment by crazymonkyyy** [comment]
stdio.remove, mutation.remove

**Reply by ntrel** [reply]
`std.algorithm.remove` should be renamed `removeIndex` (https://github.com/dlang/phobos/pull/6154#issuecomment-364699213).

> stdio.remove

Probably you meant `std.file.remove`. Possibly it could be renamed `delete` if that stopped being a keyword in the next edition. That would actually be a lot clearer than just seeing `remove(someString)`.

**Comment by crazymonkyyy** [comment]
`random.randomShuffle`

**Comment by pbackus** [comment]
`std.typecons.Nullable` should be renamed to `Optional` or `Maybe`.

We get a steady stream of users in the forums and on Discord asking why `Nullable!T` can't be assigned or compared with `null`, why `Nullable!(T*)` has two different "null" states, etc. The source of this confusion is the name `Nullable`.

**Reply by crazymonkyyy** [reply]
there should probably be two nullables, one thats overly correct and verbose and one that simply appends a bool to a type

**Reply by ichordev** [reply]
I would prefer the name `Opt`, as in optional; `Optional` and `Nullable` are just so long that I would never use them.

That said, I really wish we just had the Swift `?` syntax & pattern matching for optionals as a language feature...

**Reply by jmdavis** [reply]
Out of those, I'd go with `Maybe`, since it's shorter. I like the name `Nullable` in general, but I definitely like the idea of getting away from the issues related to folks wanting to do `is null` or getting confused about null arrays or pointers.

**Comment by pbackus** [comment]
`std.functional.bind` is too easily confused with [Javascript's `bind`][1] and [C++'s `std::bind`][2]. Its name should be changed to something without conflicting associations in other widely-used languages.

Some possibilities:

* `unpack` (from [Python's `*` operator][3])
* `spread` (from [Javascript's `...` operator][4])
* `splat` (from [Ruby's `*` operator][5])

Lisp and Scheme call this function `apply`, but D already has `opApply`, which associates a different meaning with that word.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind
[2]: https://en.cppreference.com/w/cpp/utility/functional/bind
[3]: https://docs.python.org/3/reference/expressions.html#expression-lists
[4]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
[5]: https://en.wikibooks.org/wiki/Ruby_Programming/Syntax/Method_Calls#Variable_Length_Argument_List,_Asterisk_Operator

**Reply by ichordev** [reply]
`unpack` has my vote there.

**Comment by schveiguy** [comment]
I'd also nominate `walkLength` for the axe. We have `std.algorithm.searching.count`, which does the same thing when not provided any target to search for, and is way way more intuitive.

**Reply by ntrel** [reply]
Makes sense, though `walkLength` takes an optional `upTo` argument.
https://dlang.org/phobos/std_range_primitives.html#walkLength

This works though, and is clearer: `InfiniteRange().take(4).count`.

**Comment by SRKX0A** [comment]
One addition that I think might be useful would be to change ```std.functional.not``` to ```negate``` instead. Would be related to allowing keyword forms like ```and```, ```or```, and ```not```, a la this post: https://forum.dlang.org/thread/hlxedfavwmivxllxmzuw@forum.dlang.org.

**Comment by Ogi-kun** [comment]
I think that `std.compiler.name` should be either renamed to `compilerName`, or redefined as a static member of `std.compiler.Compiler` struct. Same with other variables in this module.

**Reply by Ogi-kun** [reply]
Also, `version_major`/`version_minor` and  `D_major`/`D_major` from the same module don’t follow D naming conventions. Should be renamed to `versionMajor`/`versionMinor` and `dMajor`/`dMinor`. (Or maybe `dlangMajor`/`dlangMinor`?)

**Comment by pbackus** [comment]
`Variant` should be renamed to `Any`.

**Reply by ichordev** [reply]
I even forgot that’s what ‘variant’ was for because of the confusing name. I was looking for something like that!

**Comment by dukc** [comment]
From `std.bitmanip`, there's no reason for the "native" words in the endian functions. `aFloat.toLittleEndian` and `byteQuad.littleEndianTo!float` would work just as well. I know Átila already hates the present names and doesn't need convincing but putting it here in case this thread will be used as a sort of memo list.

**Comment by ArthaTi** [comment]
`std.conv.text` is too general for me. It's not a verb, but a noun that may easily be confused with a struct member. `concatenate` would be more descriptive of what it does, but maybe a bit awkward.

**Reply by rikkimax** [reply]
It's not concatenation either. It converts what you pass into it to a string also.

**Reply by ArthaTi** [reply]
`composeText`, perhaps.

**Reply by ichordev** [reply]
I think `text` is really fine. It's short enough that it never gets in the way, making it more practical.

**Reply by pbackus** [reply]
Re: verb vs noun, it's quite normal for functions to be named for what they return (a noun) instead of what they do (a verb). For example, we write `sqrt(x)`, not `calculateSqrt(x)`.

**Reply by ArthaTi** [reply]
True, although I do think it's still too general and it's easy to confuse. E.g. is `message.text` a field of `message` or will it just call `message.toString`?

**Reply by andrey-zherikov** [reply]
May be `toText`?

**Reply by atilaneves** [reply]
I think `text` is not just fine but pretty good.

**Reply by rikkimax** [reply]
I'm happy with ``text`` also.

But if we had to change it I'd go for ``asText`` or ``asString``.

I use ``asX`` to mean some sort of conversion with memory allocation taking place into another form.

**Reply by ArthaTi** [reply]
That would be good. Note that at least in case of `std.uni`, `asUpperCase` is the non-allocating, range variant, and `toUpper` is the allocating one. I think `toText` would work well.

**Comment by ntrel** [comment]
[`phobos.sys.meta` notes say](https://github.com/dlang/PhobosV3-Design/blob/8a67d30d089235c66469db9b6a80ada62914b91d/module_notes/phobos/sys/meta.md#unique_) NoDuplicates has:
> been renamed to be more in line with std.algorithm's uniq, which will probably be renamed to unique due to Adam Wilson being against truncated names

I think renaming NoDuplicates to Unique is OK, but the reasoning for `uniq` is not. `uniq` only removes consecutive duplicates, unlike Unique. So it would definitely be confusing & bug-prone if `uniq` was called `unique`.

**Reply by ichordev** [reply]
I hate long names but what about `sequentialUnique`?

**Reply by pbackus** [reply]
IMO a descriptive name for this should include the word "repeat." Some ideas:

* `skipRepeats`
* `mergeRepeated`
